Propagate default device settings  to internally created "reduce" tasks

vsilaev opened the pull request:
Fix for the issue https://github.com/beehive-lab/TornadoVM/issues/64 -- "reduce" tests fail when default device is altered

#### Description

Previous PR https://github.com/beehive-lab/TornadoVM/pull/61 to provide vm-wide default driver/device settings for the application is incomplete. 

#### Problem description

Code in PR61 doesn't take on account "reduction" tasks created by Tornado automatically. In effect, "reduction" tasks get default (`0:0`) device while "user-defined" tasks follow the VM-wide settings for driver/device. This leads to the error while Tornado will try to execute 2 interrelated tasks on different devices.

#### Backend/s tested

OpenCL

#### OS tested

 Windows 10, 
3 OpenCL devices - NVIDIA GPU, AMD GPU, Intel CPU (listed in this order by `tornado --devices`)

#### Did you check on FPGAs?

The code was not tested on FPGAs, it looks like `uk.ac.manchester.tornado.runtime.tasks.ReduceTaskSchedule #inspectBinariesFPGA` must be adjusted as well.

#### How to test the new patch?

You need at least 2 OpenCL devices. After compilation try to run tests that involves reduction with overridden default device, for ex:
```
tornado -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=True --printBytecodes -Dtornado.driver=0 -Dtornado.device=1 uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.reductions.TestReductionsLong
```

jjfumero commented:
All PRs are pushed to `develop`. 
Additionally, we have a new template to follow for each PR. 
https://github.com/beehive-lab/TornadoVM/blob/master/.github/pull_request_template.md

vsilaev commented:
FPGA-related code is addressed as well

vsilaev commented:
I'd like to summarize changes to avoid them looking too scary:
1. Primary affected classes are `TornadoTaskSchedule` and `ReduceTaskSchedule`
2. Explicit management of Java threads is removed and replaced with `Future` / `CompletableFuture` + `ThreadPoolExecutor` from `TornadoCoreRuntime` - btw, it's modified to use daemon threads and do not prevent JVM from exiting.
3. I minimized the spread of DEFAULT_DRIVER settings across code, now only `executeWithXXX` methods specifies what driver to use.
4. I noticed that you are using `System.setProperty to pass per-task settings in both dynamic code generation scenario and reductions scenario - I've overloaded a pair of API methods to let specify settings programmatically and avoid polluting global `System.properties` with strings.
5. On my side all tests are passed with these changes.

jjfumero commented:
Thank you @vsilaev . We will look into this in the following days. Hopefully by Monday.

jjfumero reviewed (changes requested):
Reductions are broken. If the build is only for one backend, we get the following errors:

```bash
$ make test-hdgraphics
tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats

Test: class uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats
	Running test: testSumFloats              ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloats
	Running test: testSumFloatsConstant      ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloatsConstant
	Running test: testSumFloatsLarge         ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloatsLarge
	Running test: testSumFloats2             ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloats3
	Running test: testSumFloats3             ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloats4
	Running test: testComputeAverage         ................  [PASS] 
	Running test: testMultFloats             ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 multiplyFloats
	Running test: testComputePi              ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 computePi
	Running test: testMaxReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation
	Running test: testMaxReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation2
	Running test: testMinReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation
	Running test: testMinReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation2
Test ran: 12, Failed: 11


Total Time (s): 2.1387360096
tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsDoubles

Test: class uk.ac.manchester.tornado.unittests.reductions.TestReductionsDoubles
	Running test: testSumDoubles             ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddDoubles
	Running test: testSumWithFunctionCall    ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionWithFunctionCall
	Running test: testSumDoublesLarge        ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddDoublesLarge
	Running test: testSumDoubles2            ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddDoubles2
	Running test: testSumDoubles3            ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddDoubles3
	Running test: testSumDoubles4            ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddDoubles4
	Running test: testMultDoubles            ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 multiplyDoubles
	Running test: testMaxReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation
	Running test: testMinReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation
	Running test: testRemoveOutliers         ................  [NOT SUPPORTED] 
	Running test: testMultipleReductions     ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 prepareTornadoSumForMeanComputation
	Running test: testMultipleReductions2    ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 prepareTornadoSumForMeanComputation
	Running test: testMultipleReductions3    ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 prepareTornadoSum
	Running test: testMaxReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation2
	Running test: testMinReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation2
Test ran: 14, Failed: 14


Total Time (s): 1.91749382019
tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsIntegers

Test: class uk.ac.manchester.tornado.unittests.reductions.TestReductionsIntegers
	Running test: testReductionAnnotationCPUSimple ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAnnotation
	Running test: testReductionIntsLarge     ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAnnotationLarge
	Running test: testReductionAnnotation    ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAnnotation2
	Running test: testReductionConstant      ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAnnotationConstant
	Running test: testMultiplicationReduction ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 multReductionAnnotation
	Running test: testMaxReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation
	Running test: testMinReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation
	Running test: testSequentialReduction    ................  [PASS] 
	Running test: testReduction01            ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reduction01
	Running test: testMapReduce              ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 map01
	Running test: testMapReduce3             ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 map02
	Running test: testSumInts2               ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddInts2
	Running test: testSumInts3               ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddInts3
	Running test: testMaxReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation2
	Running test: testMinReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation2
Test ran: 15, Failed: 14


Total Time (s): 2.35994911194
tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsLong

Test: class uk.ac.manchester.tornado.unittests.reductions.TestReductionsLong
	Running test: testReductionSum           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAnnotation
	Running test: testMultiplicationReduction ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 multReductionAnnotation
	Running test: testMaxReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation
	Running test: testMinReduction           ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation
	Running test: testMaxReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation2
	Running test: testMinReduction2          ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation2
Test ran: 6, Failed: 6


Total Time (s): 1.81048798561
tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsAutomatic

Test: class uk.ac.manchester.tornado.unittests.reductions.TestReductionsAutomatic
	Running test: testIrregularSize01        ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 test
	Running test: testIrregularSize02        ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 testFloat
	Running test: testIrregularSize03        ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 testFloat
	Running test: testIrregularSize04        ................  [FAILED] 
		\_[REASON] failed guarantee: No sketch available for 0:0 testDouble
Test ran: 4, Failed: 4


Total Time (s): 2.22414708138
```

jjfumero commented:
We have an API call to change the device by using `taskSchedule.mapAllTo(device)`. Why not use this call in the reductions directly? 

https://github.com/beehive-lab/TornadoVM/blob/master/tornado-api/src/main/java/uk/ac/manchester/tornado/api/TaskSchedule.java#L234-L237

Also, it looks to me the hybrid reductions are removed?

jjfumero commented on drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/runtime/OCLCompilableTask.java:
```diff
@@ -48,7 +48,11 @@
+    }
+    
+    public OCLCompilableTask(ScheduleMetaData meta, String id, Map<String, Object> properties, Method method, Object thisObject, Object... args) {
+        super(meta, id, properties, method, thisObject, args);
```
What are these properties for?

vsilaev replied:
`OCLCompilableTask extends CompilableTask` so these properties are programmatic properties passed via CompilableTask API to the superclass. Btw, `OCLCompilableTask ` is never used.

jjfumero commented on drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/tests/TestOpenCLTornadoCompiler.java:
```diff
@@ -66,6 +66,7 @@ public static void main(String[] args) {
         OCLBackend backend = tornadoRuntime.getDriver(OCLDriver.class).getDefaultBackend();
         ScheduleMetaData scheduleMeta = new ScheduleMetaData("oclbackend");
         TaskMetaData meta = new TaskMetaData(scheduleMeta, "saxpy");
+        meta.setDevice(OpenCL.defaultDevice());
```
Can you select the device that the user passes?

vsilaev replied:
Please explain. User can pass here default device via system properties `tornado.opencl.platform` and `tornado.opencl.device` -- no other app-specific command line arguments are used.

jjfumero commented:
We have an example also about how to change the default device here:

https://github.com/beehive-lab/TornadoVM/blob/master/unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java#L74

vsilaev commented on drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/OCLProgram.java:
```diff
@@ -104,7 +104,10 @@ public String getBuildLog(long deviceId) {
         buffer.clear();
-
+        if (Thread.currentThread().isInterrupted()) {
+            // Prevent ACCESS_VIOLATION in AMD devices
```
Fixed

vsilaev commented:
> We have an API call to change the device by using `taskSchedule.mapAllTo(device)`. Why not use this call in the reductions directly?
> 
> https://github.com/beehive-lab/TornadoVM/blob/master/tornado-api/src/main/java/uk/ac/manchester/tornado/api/TaskSchedule.java#L234-L237
> 
> Also, it looks to me the hybrid reductions are removed?

Using just `TaskSchedule.mapAllTo(...)` doesn't handle the case for  setting `GlobalAndLocalDimensionsFPGA` in `ReduceTaskSchedule` -- you are using `System.setProperty(...)` in this case. Otherwise it should be possible. I can revert code to left old "teleport" via `System.setProperty()` just for this and use `mapAllTo(...)` to set device.

Should I proceed?

vsilaev commented:
> Reductions are broken. If the build is only for one backend, we get the following errors:
> 
> ```shell
> $ make test-hdgraphics
> tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats
> 
> Test: class uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats
> 	Running test: testSumFloats              ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloats
> 	Running test: testSumFloatsConstant      ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloatsConstant
> 	Running test: testSumFloatsLarge         ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloatsLarge
> 	Running test: testSumFloats2             ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloats3
> 	Running test: testSumFloats3             ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 reductionAddFloats4
> 	Running test: testComputeAverage         ................  [PASS] 
> 	Running test: testMultFloats             ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 multiplyFloats
> 	Running test: testComputePi              ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 computePi
> 	Running test: testMaxReduction           ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation
> 	Running test: testMaxReduction2          ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 maxReductionAnnotation2
> 	Running test: testMinReduction           ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation
> 	Running test: testMinReduction2          ................  [FAILED] 
> 		\_[REASON] failed guarantee: No sketch available for 0:0 minReductionAnnotation2
> Test ran: 12, Failed: 11
> 
It's VERY interesting issue. The roots are in `TornadoVM.compileTaskFromBytecodeToBinary` where you overwrite task-specific device with context-specific one (coming from TaskSchedule).
For ex, the following test will run ok with the following command-line changes:
 ```bash
tornado-test.py -V -J"-Ds0.t0.device=0:2 -Ds0.t1.device=0:2 -Ds0.device=0:2" uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats
```

Will try to fix this method...

jjfumero commented:
> > We have an API call to change the device by using `taskSchedule.mapAllTo(device)`. Why not use this call in the reductions directly?
> > https://github.com/beehive-lab/TornadoVM/blob/master/tornado-api/src/main/java/uk/ac/manchester/tornado/api/TaskSchedule.java#L234-L237
> > Also, it looks to me the hybrid reductions are removed?
> 
> Using just `TaskSchedule.mapAllTo(...)` doesn't handle the case for setting `GlobalAndLocalDimensionsFPGA` in `ReduceTaskSchedule` -- you are using `System.setProperty(...)` in this case. Otherwise it should be possible. I can revert code to left old "teleport" via `System.setProperty()` just for this and use `mapAllTo(...)` to set device.
> 
> Should I proceed?

I understand it is not ideal, but the code will be much simpler. Internally we are changing how to access the devices,, so for us, it will be easy to accommodate those changes from this. So yes, please, you can proceed this way.

vsilaev commented:
> > > We have an API call to change the device by using `taskSchedule.mapAllTo(device)`. Why not use this call in the reductions directly?
> > > https://github.com/beehive-lab/TornadoVM/blob/master/tornado-api/src/main/java/uk/ac/manchester/tornado/api/TaskSchedule.java#L234-L237
> > > Also, it looks to me the hybrid reductions are removed?
> > 
> > 
> > Using just `TaskSchedule.mapAllTo(...)` doesn't handle the case for setting `GlobalAndLocalDimensionsFPGA` in `ReduceTaskSchedule` -- you are using `System.setProperty(...)` in this case. Otherwise it should be possible. I can revert code to left old "teleport" via `System.setProperty()` just for this and use `mapAllTo(...)` to set device.
> > Should I proceed?
> 
> I understand it is not ideal, but the code will be much simpler. Internally we are changing how to access the devices,, so for us, it will be easy to accommodate those changes from this. So yes, please, you can proceed this way.

Unfortunately, after thorough review I can say that this is incorrect to use `TaskSchedule.mapAllTo(...)` - just because of `all` in the name of the method, that correctly reflects that all tasks in schedule will get same device, but in this case it's necessary to set one device per concrete task. Looks like it's necessary to either left properties or introduce `TaskSchedule.setDeviceForTask` that will compliment existing `TaskSchedule.getDeviceForTask`. And `SchedulableTask` along with 2 subclasses `CompilableTask` and  `BuiltinTask` already have necessary mathod `mapTo(device)`. What do you think?

vsilaev commented:
Now all failed tests reported should pass

vsilaev commented:
@jjfumero , I have no feedback from you - but nevertheless, I modified the code and remove any `*.properties()` API extensions and introduced `TaskSchedule.setDeviceForTask` that pairs `TaskSchedule.getDeviceForTask`. Also this allowed to revert back a lot of changes compared to the baseline of the `develop` branch.

vsilaev commented:
Also I've reverted changes to TornadoVM to let reduction tests with alt. device working. Now it works without changes while correspondence between execution contexts and tasks' devices are maintained correctly.

vsilaev commented:
I'm about to "celebrate" 3 months of keeping this PR up to date. Do you still have plans to review and apply it?
Also, I have a pending prototype of async (non-blocking) execution of OpenCL queue-s -- and these changes depend on the PR in question. (CUDA is planned as well).

Please confirm if you plan to apply the PR and you are interesting in the next work from me, or it's simpler to start my own fork.

jjfumero commented:
Hi @vsilaev. 
  thanks for your interest, and the reminder ;-) 

Concerning the current PR: the changes proposed mix refactoring with a new functionality. I suggest to split it. Moreover, we believe there is an easier way to change device for reductions. Note that TornadoVM builds an abstraction of the OpenCL and PTX devices. So bypassing can create some issues internally. I suggest agreeing in a design before finalizing the PR. Could you please open a shared doc with your proposal? 

> Also, I have a pending prototype of async (non-blocking) execution of OpenCL queue-s -- and these changes depend on the PR in question. (CUDA is planned as well).

For the async feature, please, open a proposal in a shared document (e.g., Google doc). So we can iterate and co-design the changes. This is important for us, since we have ongoing European projects in which are implementing new features. Having a document will help the TornadoVM team to design new features with the new proposals in mind. 

Looking forward to your proposals.
Juan

vsilaev commented:
> Concerning the current PR: the changes proposed mix refactoring with a new functionality.

Could you point where exactly? The API changes are minimal and just exposes fields to access task metadata by id. This is necessary for impl. Nothing besides this.

> Note that TornadoVM builds an abstraction of the OpenCL and PTX devices. So bypassing can create some issues internally.

In this PR nothing is bypassed. PTX vs OpenCL is related for "async" and only due to fact that I have less experience with CUDA.

jjfumero commented:
> Could you point where exactly? The API changes are minimal and just exposes fields to access task metadata by id. This is necessary for impl. Nothing besides this.

Here's an example. Changes like this are for refactoring. This does not have anything to do with the implementation of the device index for the reduction. 

https://github.com/beehive-lab/TornadoVM/pull/65/files#diff-e734b0a600c8c9d0a7ee3de55cd7d1ff4b9570f8868247828d0e8602e340e49cR1266-R1297 


There is already a way to select the device (`-D<shedulename>.<taskNAme>.device=<driver:deviceIndex>`). TornadoVM builds an abstraction for all devices. You can take advantage of this abstraction and select this from the TornadoVM driver. There is a driver per backend (OpenCL and PTX).

vsilaev commented:
Well, I disagree that such config-driven settings are better than API-driven. And trashing. System.properties with numerous settings per task is not the best option. Plus you are not taking on account that this was proposed to supply VM-wide defaults rather than per-task settings - otherwise it's a tedious task to implement smth. like app->device affinity (app uses specified device).

Anyway, i'm going to ~~build my own theme park~~ maintain my own fork... with blackjack and async.

Closing PR.

jjfumero commented:
The suggested changes do not require system properties. It will take the device from the information stored in the driver implementation of TornadoVM. 

Again this PR mixes new functionality with unrelated refactor. What we suggested was splitting the PR and follow the guidelines. 
Furthermore, we haven't got any proposals.
